lib/spack/spack/util/path.py

```diff
@@ -246,23 +264,31 @@ def canonicalize_path(path, default_wd=None):
     Returns:
         (str): An absolute path with path variable substitution
     """
-    # Get file in which path was written in case we need to make it absolute
-    # relative to that path.
-    filename = None
-    if isinstance(path, syaml.syaml_str):
-        filename = os.path.dirname(path._start_mark.name)
-        assert path._start_mark.name == path._end_mark.name
-
-    path = substitute_path_variables(path)
-    if not os.path.isabs(path):
-        if filename:
-            path = os.path.join(filename, path)
-        else:
-            base = default_wd or os.getcwd()
-            path = os.path.join(base, path)
-            tty.debug("Using working directory %s as base for abspath" % base)
-
-    return os.path.normpath(path)
+
+    def _canonicalize_path(path, default_wd=None):
+        # Get file in which path was written in case we need to make it absolute
+        # relative to that path.
+        filename = None
+        if isinstance(path, syaml.syaml_str):
+            filename = os.path.dirname(path._start_mark.name)
+            assert path._start_mark.name == path._end_mark.name
+
+        path = substitute_path_variables(path)
+        if not os.path.isabs(path):
+            if filename:
+                path = os.path.join(filename, path)
+            else:
+                base = default_wd or os.getcwd()
+                path = os.path.join(base, path)
+                tty.debug("Using working directory %s as base for abspath" % base)
+
+        return os.path.normpath(path)
+
+    if env:
+        with env:
+            return _canonicalize_path(path, default_wd)
+    else:
+        return _canonicalize_path(path, default_wd)
```

Comment:
```suggestion
    with env:
        return _canonicalize_path(path, default_wd)
```
`env` cannot be None as currently written

Reply:
Considering this, probably best not to use an internal method and just do:

```
def canonicalize_path(...):
    with env:
        filename = None
        ...

Reply:
With the context manager I was suggesting you'd want something like this:

```suggestion
    # If a dictionary was provided, use it, otherwise use the frozen environment
    # as context for substituting environment variables.
    context = spack.util.lang.nullcontext()
    replacement = env if env is not None else _frozen_environment
    if replacement is not None:
        context = replace_env(replacement)
        
    with context:
        return _canonicalize_path(path, default_wd)
```

And then we could call freeze_environment() to freeze the environment before the build starts as in my earlier example, but still keep this function flexible like @georgiastuart wants.

Reply:
This makes sense to me! I'll commit the changes tomorrow and remove the old warning wording.